File: Leetcode/valid_parentheses.py
# LeetCode task: solution -> https://leetcode.com/problems/valid-parentheses/

class Solution:
    def isValid(self, s: str) -> bool:
        """
        Determines whether a string containing brackets is valid.

        Args:
            s (str): The input string containing brackets.

        Returns:
            bool: True if the string is valid, False otherwise.
        """
        stack = []  # Initialize an empty stack to track opening brackets
        brackets = {'(': ')', '{': '}', '[': ']'}  # Mapping of opening brackets to closing brackets

        for char in s:
            if char in brackets:  # If the character is an opening bracket
                stack.append(char)  # Push it onto the stack
            elif char in brackets.values():  # If the character is a closing bracket
                if not stack:  # If the stack is empty, there is no corresponding opening bracket
                    return False
                opening_bracket = stack.pop()  # Pop the top element from the stack
                if brackets[opening_bracket] != char:  # Check if the closing bracket matches the opening bracket
                    return False

        return len(stack) == 0  # If the stack is empty, all opening brackets were closed properly

File: Go programming/Sample/valid_parentheses.go
// Leetcode task: solution -> https://leetcode.com/problems/valid-parentheses/

package main

import "fmt"

func isValid(s string) bool {
	stack := []rune{} // Initialize an empty stack to track opening brackets
	brackets := map[rune]rune{'(': ')', '{': '}', '[': ']'} // Mapping of opening brackets to closing brackets

	for _, char := range s {
		if _, ok := brackets[char]; ok { // If the character is an opening bracket
			stack = append(stack, char) // Push it onto the stack
		} else if _, ok := brackets[char]; ok { // If the character is a closing bracket
			if len(stack) == 0 { // If the stack is empty, there is no corresponding opening bracket
				return false
			}
			openingBracket := stack[len(stack)-1] // Get the top element from the stack
			stack = stack[:len(stack)-1]          // Pop the top element from the stack
			if brackets[openingBracket] != char { // Check if the closing bracket matches the opening bracket
				return false
			}
		}
	}

	return len(stack) == 0 // If the stack is empty, all opening brackets were closed properly
}

func main() {
	fmt.Println(isValid("()"))       // Output: true
	fmt.Println(isValid("()[]{}"))   // Output: true
	fmt.Println(isValid("(]"))       // Output: false
}
